File: lint/check_mod.py
#!/usr/bin/env python3
"""Fake Game mod 校验工具。

用法：python lint/check_mod.py <mod路径>

校验 mod 的 manifest.json 和 content.json 是否符合 specs/mod-spec.md。
结果打印到标准输出；通过返回 0，失败返回 1。
"""
import json
import re
import sys
from pathlib import Path

ITEM_TYPES = {"weapon", "armor", "consumable"}
NAME_RE = re.compile(r"^[a-z][a-z0-9_]*$")
SEMVER_RE = re.compile(r"^\d+\.\d+\.\d+$")


def main(mod_dir: Path) -> int:
    if not mod_dir.is_dir():
        print(f"FAIL: {mod_dir} 不是目录")
        return 1

    errors = []

    # manifest.json
    manifest_path = mod_dir / "manifest.json"
    if not manifest_path.is_file():
        errors.append("缺少 manifest.json")
    else:
        try:
            manifest = json.loads(manifest_path.read_text(encoding="utf-8"))
        except Exception as exc:
            errors.append(f"manifest.json 不是合法 JSON: {exc}")
        else:
            for key in ("name", "version", "description", "author"):
                if not manifest.get(key):
                    errors.append(f"manifest.{key} 缺失或为空")
            if not SEMVER_RE.match(manifest.get("version", "")):
                errors.append(f"manifest.version 不是 semver: {manifest.get('version')!r}")
            if manifest.get("name") != mod_dir.name:
                errors.append(f"manifest.name ({manifest.get('name')!r}) 与目录名 ({mod_dir.name}) 不一致")
            if not NAME_RE.match(manifest.get("name", "")):
                errors.append(f"manifest.name 不符合 lower_snake_case: {manifest.get('name')!r}")

    # content.json
    content_path = mod_dir / "content.json"
    if not content_path.is_file():
        errors.append("缺少 content.json")
    else:
        try:
            content = json.loads(content_path.read_text(encoding="utf-8"))
        except Exception as exc:
            errors.append(f"content.json 不是合法 JSON: {exc}")
        else:
            items = content.get("items")
            if not isinstance(items, list) or not items:
                errors.append("content.items 必须是非空数组")
            else:
                seen = set()
                for i, item in enumerate(items):
                    prefix = f"items[{i}]"
                    if not isinstance(item, dict):
                        errors.append(f"{prefix} 必须是对象")
                        continue
                    for key in ("id", "name", "type"):
                        if not item.get(key):
                            errors.append(f"{prefix}.{key} 缺失或为空")
                    itype = item.get("type")
                    if itype not in ITEM_TYPES:
                        errors.append(f"{prefix}.type 必须是 {sorted(ITEM_TYPES)} 之一，得到 {itype!r}")
                    iid = item.get("id", "")
                    if not NAME_RE.match(iid):
                        errors.append(f"{prefix}.id 不符合 lower_snake_case: {iid!r}")
                    if iid in seen:
                        errors.append(f"{prefix}.id 重复: {iid!r}")
                    seen.add(iid)

    if errors:
        for error in errors:
            print(f"FAIL: {error}")
        return 1

    print(f"PASS: {mod_dir.name} 校验通过")
    return 0


if __name__ == "__main__":
    if len(sys.argv) != 2:
        print(__doc__)
        sys.exit(2)
    sys.exit(main(Path(sys.argv[1])))

File: lint/check_mod.mjs
#!/usr/bin/env node
/**
 * Fake Game mod validator.
 *
 * Usage: node lint/check_mod.mjs <mod-path>
 *
 * Validates a mod's manifest.json and content.json against specs/mod-spec.md.
 * Prints the result to stdout. Exit codes: 0 = pass, 1 = fail, 2 = usage error.
 *
 * Contract: read-only (mod + specs), never writes back to lint/ or the
 * environment; zero third-party deps (Node built-in modules only). Messages
 * are ASCII/English to avoid Windows console encoding issues (GBK vs UTF-8).
 */
import { readFileSync, statSync } from "node:fs";
import { join, basename } from "node:path";

const ITEM_TYPES = new Set(["weapon", "armor", "consumable"]);
const NAME_RE = /^[a-z][a-z0-9_]*$/;
const SEMVER_RE = /^\d+\.\d+\.\d+$/;

function main() {
  if (process.argv.length !== 3) {
    console.log("Usage: node lint/check_mod.mjs <mod-path>");
    process.exitCode = 2;
    return;
  }

  const modDir = process.argv[2];
  let isDir = false;
  try {
    isDir = statSync(modDir).isDirectory();
  } catch {
    isDir = false;
  }
  if (!isDir) {
    console.log(`FAIL: ${modDir} is not a directory`);
    process.exitCode = 1;
    return;
  }

  const errors = [];

  // manifest.json
  const manifest = readJson(join(modDir, "manifest.json"), "manifest.json", errors);
  if (manifest !== null) {
    for (const key of ["name", "version", "description", "author"]) {
      if (!manifest[key]) errors.push(`manifest.${key} is missing or empty`);
    }
    if (!SEMVER_RE.test(manifest.version ?? "")) {
      errors.push(`manifest.version is not semver x.y.z: ${JSON.stringify(manifest.version)}`);
    }
    if (manifest.name !== basename(modDir)) {
      errors.push(
        `manifest.name (${JSON.stringify(manifest.name)}) does not match dir name (${basename(modDir)})`,
      );
    }
    if (!NAME_RE.test(manifest.name ?? "")) {
      errors.push(`manifest.name is not lower_snake_case: ${JSON.stringify(manifest.name)}`);
    }
  }

  // content.json
  const content = readJson(join(modDir, "content.json"), "content.json", errors);
  if (content !== null) {
    const items = content.items;
    if (!Array.isArray(items) || items.length === 0) {
      errors.push("content.items must be a non-empty array");
    } else {
      const seen = new Set();
      items.forEach((item, i) => {
        const prefix = `items[${i}]`;
        if (typeof item !== "object" || item === null || Array.isArray(item)) {
          errors.push(`${prefix} must be an object`);
          return;
        }
        for (const key of ["id", "name", "type"]) {
          if (!item[key]) errors.push(`${prefix}.${key} is missing or empty`);
        }
        if (!ITEM_TYPES.has(item.type)) {
          errors.push(
            `${prefix}.type must be one of [${[...ITEM_TYPES].sort().join(", ")}], got ${JSON.stringify(item.type)}`,
          );
        }
        const id = item.id ?? "";
        if (!NAME_RE.test(id)) errors.push(`${prefix}.id is not lower_snake_case: ${JSON.stringify(id)}`);
        if (seen.has(id)) errors.push(`${prefix}.id is duplicated: ${JSON.stringify(id)}`);
        seen.add(id);
      });
    }
  }

  if (errors.length) {
    for (const error of errors) console.log(`FAIL: ${error}`);
    process.exitCode = 1;
    return;
  }
  console.log(`PASS: ${basename(modDir)} is valid`);
  process.exitCode = 0;
}

function readJson(filePath, label, errors) {
  try {
    return JSON.parse(readFileSync(filePath, "utf8"));
  } catch (err) {
    if (err?.code === "ENOENT") errors.push(`missing ${label}`);
    else errors.push(`${label} is not valid JSON: ${err.message}`);
    return null;
  }
}

main();
